improve resulting displacement vectors

carlocastoldi opened the pull request:
## Description

**What is this PR**

- [ ] Bug fix
- [ ] Addition of a new feature
- [x] Other

**Why is this PR needed?**
Currently, `compute_displacement` computes backward displacement vectors that are not easy to interpret.

**What does this PR do?**
As a solution, I propose to add two twin functions: `compute_forward_displacement` and `compute_backward_displacement`, whose results are vectors whose origin point is always the position in the associated frame.
I also updated the docstrings, the example and the tests accordingly.

Additionally, I also added an example of use of `compute_path_length` function.

## References

#136

## How has this PR been tested?

I have tested it by running the `compute_kinematics` example and checking whether the resulting vectors follow the direction of points.
Finally, i made sure unit and integration tests were correctly adapted and passing.

## Is this a breaking change?

Yes, it deprecates `compute_displacement` in favour of `compute_forward_displacement` and `compute_backward_displacement`.
See an example below:

```python
# Instead of:
from movement.kinematics import compute_displacement
bwd_inverted_displacement = compute_displacement(ds.position)

# Use:
from movement.kinematics import compute_backward_displacement
bwd_inverted_displacement = -compute_backward_displacement(ds.position)
```


## Does this PR require an update to the documentation?

The documentation should be re-generated to reflect the changes in the doctrings

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [x] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

sfmig reviewed (changes requested):
Very nice contribution @carlocastoldi! 🎉 

I took the chance to update the example a bit, and also included the backward displacement vector computation (that way in the API docs the example is linked for both functions, see [here](https://movement.neuroinformatics.dev/api/movement.kinematics.compute_acceleration.html#examples-using-compute-acceleration) for example). I pushed the changes directly to your branch - hope that is not too confusing!

I have a few small comments I left in line. For some of the more meaty ones I'm curious about your thoughts. 

Apart from that, it would be very useful for us if you can document the deprecation of `compute_displacement`. That means:
- [ ] writing a clear old-vs-new syntax in the description of this PR. This helps Niko write a small "migration guide" when we make a release (see the "Breaking changes" section [here](https://movement.neuroinformatics.dev/blog/movement-v0_0_21.html)).
- [ ] add a `DeprecationWarning` under a `compute_displacement` function, which we will remove after a few releases. We did this for the old implementation of `median_filter` , you can see the example [here](https://github.com/neuroinformatics-unit/movement/pull/455/commits/43914fba1c125e6f8f919d1a1d2fb10a92ab0e31). 
We are planning to formalise this a bit more (see #663) but we are not there yet.

Thanks for the great work!

carlocastoldi commented on movement/kinematics/kinematics.py:
```diff
@@ -60,15 +60,16 @@ def compute_time_derivative(data: xr.DataArray, order: int) -> xr.DataArray:
-def compute_displacement(data: xr.DataArray) -> xr.DataArray:
-    """Compute displacement array in cartesian coordinates.
+def compute_forward_displacement(data: xr.DataArray) -> xr.DataArray:
+    """Compute forward displacement array in cartesian coordinates.
```
I've changed it in other docstrings too!

carlocastoldi commented on tests/test_integration/test_kinematics_vector_transform.py:
```diff
@@ -15,20 +15,45 @@
+                        # However, current test implementation forces phi=0.
                     ],
-                ),  # Individual 0, rho=sqrt(2), phi=45deg
+                ),  # Individual 0, rho=√2, phi=45deg=π/4=atan(1)
```
Sure!

carlocastoldi commented on tests/test_integration/test_kinematics_vector_transform.py:
```diff
@@ -15,20 +15,45 @@
+                        np.tile([math.sqrt(2), math.pi / 4], (9, 1)),
                         np.zeros((1, 2)),
-                        np.tile([math.sqrt(2), math.atan(1)], (9, 1)),
+                        # at time t=-1, the forward displacement cartesian
+                        # vector is (x=0,y=0).
+                        # This means that while rho=0, phi can be whatever.
+                        # However, current test implementation forces phi=0.
```
Yeah, I understand your point. This is an integration test between two components that should be deterministic in the results they have.
I agree that it's probably better to just mention/test this in `cart2pol` docstrings.

The integration test, then, can test according to the assumptions on `phi` indirectly derived from numpy's arctan2 table because that's what `cart2pol` promises to return.

Should I modify cart2pol in this PR too?

sfmig replied:
I have added this to cart2pol now, thanks!

carlocastoldi commented on tests/test_unit/test_kinematics/test_kinematics.py:
```diff
@@ -8,19 +8,31 @@
 class TestComputeKinematics:
     """Test ``compute_[kinematic_variable]`` with valid and invalid inputs."""
 
+    forward_displacement = [
+        np.vstack([np.ones((9, 2)), np.zeros((1, 2))]),
+        np.multiply(
+            np.vstack([np.ones((9, 2)), np.zeros((1, 2))]),
+            np.array([1, -1]),
+        ),
+    ]  # [Individual 0, Individual 1]
```
yes, the reason I went for an explicit definition of the `backward_displacement` results is because I thought it was clearer in showing that there are exactly two tests derived from `forward_displacement`.

If you prefer the list comprehension, tho, I also understand! Perhaps I could just change the name of the variables to the plurals: `forward_displacement` -> `forward_displacements` and `backward_displacement` -> `backward_displacements`

carlocastoldi commented on tests/test_integration/test_kinematics_vector_transform.py:
```diff
@@ -15,20 +15,45 @@
+                        np.zeros((1, 2)),
+                        np.tile([math.sqrt(2), math.pi * 3 / 4], (9, 1)),
+                    ]
+                ),  # Individual 1, rho=√2, phi=135deg=π3/4
```
beautiful!
I simplified it even further by rolling the forward vectors here too, in order to have the backwards'

carlocastoldi commented:
Thank you for the review and for updating the examples. They look good!

I have:
* reintroduced `compute_displacement`, deprecated it and tested for its deprecation
* changed `cart2pol` to be more intuitive when the cartesian vector is of length=0
* extracted `_compute_forward_displacement` function to be used by both `compute_forward_displacement` and `compute_backward_displacement`
* reduced the clutter of parametrisation of integration tests
* changed usage in integration tests of `math.atan` in favour of `math.pi` notation
* updated the PR with a clear example of old-vs-new syntax
* updated docstrings with forward/backward displacement specifics and capitalised "Cartesian"

sfmig commented on movement/utils/vector.py:
```diff
@@ -119,6 +119,7 @@ def cart2pol(data: xr.DataArray) -> xr.DataArray:
         data.sel(space="y"),
         data.sel(space="x"),
     )
+    phi = phi.where(rho != 0, 0)
```
For reference, since these are floats I replaced this with an is-close approach (see e.g. [here](https://www.geeksforgeeks.org/python/comparing-floating-points-number-for-almost-equality-in-python/) for more details)

sfmig commented:
Thank you @carlocastoldi! You did a fantastic job, it was a pleasure to work together on this. 

I am very glad the displacement vector is in order, now we can all sleep at night 😂 😂 

If you fancy contributing to other `movement` or NIU bits at any point do reach out, I'd be more than happy to work on something together again.
